src/TestFramework/PhpUnit/Coverage/IndexXmlCoverageParser.php

```diff
@@ -0,0 +1,293 @@
+    {
+        $lineCoverage = self::safeQuery($xPath, '/phpunit/project/directory[1]/totals/lines')->item(0);
+
+        if (
+            !($lineCoverage instanceof DOMElement)
+            || ($coverageCount = $lineCoverage->getAttribute('executed')) === '0'
+            || $coverageCount === ''
+        ) {
+            throw NoLineExecuted::create();
+        }
+    }
+
+    /**
+     * @param array<string, CoverageFileData> $data
+     *
+     * @throws CoverageDoesNotExistException
+     */
+    private function processCoverageFile(
+        string $relativeCoverageFilePath,
+        string $projectSource,
+        array &$data
+    ): void {
+        $xPath = self::createXPath(file_get_contents(
+            realpath($projectSource . '/' . $relativeCoverageFilePath)
+        ));
+
+        $sourceFilePath = self::retrieveSourceFilePath($xPath, $relativeCoverageFilePath, $projectSource);
+
+        $linesNode = self::safeQuery($xPath, '/phpunit/file/totals/lines')[0];
+
+        $percentage = $linesNode->getAttribute('percent');
+
+        if ($percentage === '') {
+            $data[$sourceFilePath] = new CoverageFileData();
+
+            return;
+        }
+
+        Assert::numeric($percentage);
+
+        $coveredLineNodes = self::safeQuery($xPath, '/phpunit/file/coverage/line');
+
+        if ($coveredLineNodes->length === 0) {
+            $data[$sourceFilePath] = new CoverageFileData();
+
+            return;
+        }
+
+        $coveredMethodNodes = self::safeQuery($xPath, '/phpunit/file/class/method');
+
+        if ($coveredMethodNodes->length === 0) {
+            $coveredMethodNodes = self::safeQuery($xPath, '/phpunit/file/trait/method');
+        }
+
+        $data[$sourceFilePath] = new CoverageFileData(
+            self::collectCoveredLinesData($coveredLineNodes),
+            self::collectMethodsCoverageData($coveredMethodNodes)
+        );
+    }
+
+    /**
+     * @throws CoverageDoesNotExistException
+     */
+    private static function retrieveSourceFilePath(
+        DOMXPath $xPath,
+        string $relativeCoverageFilePath,
+        string $projectSource
+    ): string {
+        $fileNode = self::safeQuery($xPath, '/phpunit/file')[0];
+
+        $fileName = $fileNode->getAttribute('name');
+        $relativeFilePath = $fileNode->getAttribute('path');
+
+        if ($relativeFilePath !== '') {
+            // The relative path is not present for old versions of PHPUnit. As a result we parse
+            // the relative path from the source file path and the XML coverage file
+            $relativeFilePath = str_replace(
+                sprintf('%s.xml', $fileName),
+                '',
+                $relativeCoverageFilePath
+            );
+        }
+
+        $path = $projectSource . '/' . ltrim($relativeFilePath, '/') . '/' . $fileName;
+        $realPath = native_realpath($path);
+
+        if ($realPath === false) {
+            throw CoverageDoesNotExistException::forFileAtPath($fileName, $path);
```

Comment:
how can this scenario happen in practice? It's not covered by the tests either

Reply:
I don't know, probably it's just for PHPStan?

Reply:
Actually can happen if the coverage is out of sync with the source code